File: pkg/nextest/nextest.go
// Package nextest runs Nexus test files (*.nex) with assert builtins.
package nextest

import (
	"fmt"
	"io"
	"os"
	"path/filepath"
	"strings"
	"time"

	"nex-lang/pkg/evaluator"
	"nex-lang/pkg/runtime"
)

// Result is the outcome of one test file.
type Result struct {
	File     string
	Passed   bool
	Duration time.Duration
	Error    string
}

// Summary aggregates a test run.
type Summary struct {
	Results []Result
	Passed  int
	Failed  int
}

// Options configures discovery and execution.
type Options struct {
	RootDir   string
	StdlibDir string
	// Paths are explicit files or directories; empty means discover under RootDir.
	Paths []string
	Out   io.Writer
}

// Run executes discovered tests and returns a summary.
func Run(opts Options) (*Summary, error) {
	if opts.Out == nil {
		opts.Out = os.Stdout
	}
	if opts.RootDir == "" {
		cwd, err := os.Getwd()
		if err != nil {
			return nil, err
		}
		opts.RootDir = cwd
	}

	files, err := discover(opts.RootDir, opts.Paths)
	if err != nil {
		return nil, err
	}
	if len(files) == 0 {
		return nil, fmt.Errorf("no test files found (looked for *_test.nex and tests/**/*.nex)")
	}

	sum := &Summary{}
	for _, file := range files {
		res := runFile(file, opts)
		sum.Results = append(sum.Results, res)
		rel := file
		if r, err := filepath.Rel(opts.RootDir, file); err == nil {
			rel = r
		}
		if res.Passed {
			sum.Passed++
			fmt.Fprintf(opts.Out, "ok   %s (%.2fms)\n", rel, float64(res.Duration.Microseconds())/1000)
		} else {
			sum.Failed++
			fmt.Fprintf(opts.Out, "FAIL %s (%.2fms)\n  %s\n", rel, float64(res.Duration.Microseconds())/1000, res.Error)
		}
	}
	fmt.Fprintf(opts.Out, "\n%d passed, %d failed\n", sum.Passed, sum.Failed)
	return sum, nil
}

func discover(root string, paths []string) ([]string, error) {
	var files []string
	seen := map[string]bool{}
	add := func(p string) {
		p = filepath.Clean(p)
		if seen[p] {
			return
		}
		seen[p] = true
		files = append(files, p)
	}

	if len(paths) == 0 {
		paths = []string{
			filepath.Join(root, "tests"),
			root,
		}
	}

	for _, p := range paths {
		if !filepath.IsAbs(p) {
			p = filepath.Join(root, p)
		}
		st, err := os.Stat(p)
		if err != nil {
			if os.IsNotExist(err) && len(paths) > 1 {
				continue
			}
			return nil, err
		}
		if !st.IsDir() {
			if strings.HasSuffix(p, ".nex") {
				add(p)
			}
			continue
		}
		err = filepath.WalkDir(p, func(path string, d os.DirEntry, err error) error {
			if err != nil {
				return err
			}
			if d.IsDir() {
				name := d.Name()
				if name == "node_modules" || name == ".git" || name == "bin" || name == "vscode-nexus" {
					return filepath.SkipDir
				}
				return nil
			}
			base := d.Name()
			rel, _ := filepath.Rel(root, path)
			inTests := strings.HasPrefix(filepath.ToSlash(rel), "tests/")
			if strings.HasSuffix(base, "_test.nex") || (inTests && strings.HasSuffix(base, ".nex")) {
				add(path)
			}
			return nil
		})
		if err != nil {
			return nil, err
		}
	}
	return files, nil
}

func runFile(file string, opts Options) Result {
	start := time.Now()
	env := evaluator.NewEnvironment()
	appDir := filepath.Dir(file)
	rt := runtime.NewWithOptions(appDir, env, runtime.Options{
		StdlibDir:  opts.StdlibDir,
		ModulesDir: filepath.Join(opts.RootDir, runtime.ModulesDir),
	})
	result := rt.LoadFile(file)
	res := Result{File: file, Duration: time.Since(start)}
	if errObj, ok := result.(*evaluator.Error); ok {
		res.Error = errObj.Message
		return res
	}
	res.Passed = true
	return res
}

File: vscode-nexus/src/nextest.ts
/**
 * Discover and run Nexus test files (*_test.nex and tests/*.nex).
 */

import * as fs from "fs";
import * as path from "path";
import { evaluate } from "./compiler/engine";
import { findStdlibDir, MODULES_DIRNAME } from "./runtime/modules";
import { isError } from "./language/values";

export interface TestResult {
  file: string;
  passed: boolean;
  durationMs: number;
  error?: string;
  output: string[];
}

export interface TestSummary {
  results: TestResult[];
  passed: number;
  failed: number;
}

export interface TestOptions {
  rootDir: string;
  paths?: string[];
  stdlibDir?: string;
  out?: (line: string) => void;
}

function discover(root: string, paths: string[]): string[] {
  const files: string[] = [];
  const seen = new Set<string>();
  const add = (p: string) => {
    const cleaned = path.normalize(p);
    if (seen.has(cleaned)) {
      return;
    }
    seen.add(cleaned);
    files.push(cleaned);
  };

  const searchRoots =
    paths.length > 0 ? paths : [path.join(root, "tests"), root];

  for (let p of searchRoots) {
    if (!path.isAbsolute(p)) {
      p = path.join(root, p);
    }
    let st: fs.Stats;
    try {
      st = fs.statSync(p);
    } catch {
      continue;
    }
    if (!st.isDirectory()) {
      if (p.endsWith(".nex")) {
        add(p);
      }
      continue;
    }
    walk(p, root, add);
  }
  return files;
}

function walk(
  dir: string,
  root: string,
  add: (p: string) => void,
): void {
  for (const entry of fs.readdirSync(dir, { withFileTypes: true })) {
    const full = path.join(dir, entry.name);
    if (entry.isDirectory()) {
      if (
        entry.name === "node_modules" ||
        entry.name === ".git" ||
        entry.name === "out" ||
        entry.name === "bin"
      ) {
        continue;
      }
      walk(full, root, add);
      continue;
    }
    const rel = path.relative(root, full).replace(/\\/g, "/");
    const inTests = rel.startsWith("tests/");
    if (
      entry.name.endsWith("_test.nex") ||
      (inTests && entry.name.endsWith(".nex"))
    ) {
      add(full);
    }
  }
}

export async function runTests(opts: TestOptions): Promise<TestSummary> {
  const log = opts.out ?? ((line: string) => {
    // eslint-disable-next-line no-console
    console.log(line);
  });
  const files = discover(opts.rootDir, opts.paths ?? []);
  if (files.length === 0) {
    throw new Error(
      "no test files found (looked for *_test.nex and tests/**/*.nex)",
    );
  }

  const stdlibDir =
    opts.stdlibDir !== undefined
      ? opts.stdlibDir
      : findStdlibDir(opts.rootDir);
  const summary: TestSummary = { results: [], passed: 0, failed: 0 };

  for (const file of files) {
    const start = Date.now();
    const source = fs.readFileSync(file, "utf8");
    const result = await evaluate(source, {
      tier: "eval",
      rootDir: opts.rootDir,
      modulesDir: path.join(opts.rootDir, MODULES_DIRNAME),
      stdlibDir,
      filePath: file,
      checkOwnership: false,
      enableEffects: false,
    });
    const durationMs = Date.now() - start;
    const rel = path.relative(opts.rootDir, file) || file;
    const failed = isError(result.value);
    const tr: TestResult = {
      file,
      passed: !failed,
      durationMs,
      error: failed ? result.value.inspect() : undefined,
      output: result.output,
    };
    summary.results.push(tr);
    if (tr.passed) {
      summary.passed++;
      log(`ok   ${rel} (${durationMs}ms)`);
    } else {
      summary.failed++;
      log(`FAIL ${rel} (${durationMs}ms)`);
      log(`  ${tr.error}`);
    }
  }

  log("");
  log(`${summary.passed} passed, ${summary.failed} failed`);
  return summary;
}
